### ISSUE TITLE
Mismatching pod IP and datastore IP during deletion

### ISSUE THREAD

**Issue by duxing** [aws/amazon-vpc-cni-k8s#2352]
This maybe a dumb question, but I'd like to understand the internals of this project better.

**What happened**:

I'm going through the verbose `aws-cni` logs of a particular node and running into a sequence I'm having difficulty explaining. skipping a few logs to reduce verbosity here but the overall flow is as following:

 - 1st log: current state is healthy
> Current warm IP stats: target: 10, total: 48, assigned: 38, available: 10, short: 0, over 0
 - 2nd log: received a delete request for a pod of mine. this is normal as well.
> Received CNI del request: xxxx
 - 3rd log: involved in processing the delete request
> UnassignPodIPv4Address: IP address pool stats: total:48, assigned 38, sandbox aws-cni/<container_id>/eth0
 - 4th log: `datastore` unassign the IP. where `10.79.169.100` is the pod IP for the pod being deleted.
> UnAssignPodIPv4Address: Unassign IP 10.79.169.100 from sandbox aws-cni/<container_id>/eth0
 - 5th log: `ipamd` issues a rpc call to cni plugin to unassign the IP from above. and cni plugin proceed with the deletion:
> Send DelNetworkReply: IPv4Addr 10.79.169.100, DeviceNumber: 1, err: <nil>
> Received del network response for pod 
> TeardownNS: addr 10.79.169.100/32, deviceNumber 1
> Delete toContainer rule for 10.79.169.100/32 
> Delete fromContainer rule for 10.79.169.100/32 in table 2
> Tear down of NS complete
 - 6th log from `ipamd` state dump, reflecting the ip unassignment
> Current warm IP stats: target: 10, total: 48, assigned: 37, available: 11, short: 0, over 1
 - 7th log: ipamd start to decrease datastore pool
> Starting to decrease Datastore pool
 - 8th log: unable to find unassigned ip, but will retry in 30s (from source code)
> Error finding unassigned IPs for ENI <eni-id>
 - 9th log: start to decrease datastore pool again (30s after 7th log):
> Starting to decrease Datastore pool
> Deleted ENI(<eni-id>)'s IP/Prefix 10.79.162.37/32 from datastore
> Trying to unassign the following IPs [10.79.162.37] from ENI <eni-id>
> Successfully freed IPs [10.79.162.37] from ENI <eni-id>

for context:
 - I'm running `v1.9.3`
 - instance type is `r5.8xlarge`: 30 ips per ENI, 8 max ENI. which can be verified from the logs I found:
> IP pool stats: total = 48, used = 37, c.maxIPsPerENI = 29
 - the subnet I used contains more than enough IPs

my question is: 
 - why pod ip being deleted is `10.79.169.100` where the ip being deleted from datastore is `10.79.162.37`? there's no other concurrent delete request in the span I post here. I'd like to know if these 2 ips are supposed to be the same.

I've checked another delete request and saw the same issue. Not sure if this is expected but I'd like to get a better understanding. Seems to me this is a state drift between ENI ips and the datastore.

Apologize if this is a dumb question. I can't figure out after following the troubleshooting steps. Thanks in advance!

**Environment**:
- Kubernetes version (use `kubectl version`): `1.14`
- CNI Version: `1.9.3`
- OS (e.g: `cat /etc/os-release`): debian 9
- Kernel (e.g. `uname -a`): `Debian 4.9.246-2 (2020-12-17) x86_64 GNU/Linux`

**Comment by duxing** [comment]
follow up on this issue. I noticed another `warn`ing log that reflects a similar issue.

```
Instancemetadatadoesnotmatchdatastore!ipPool: [
  10.79.162.143,
  10.79.173.187,
  10.79.163.21,
  10.79.165.42,
  10.79.161.141,
  10.79.166.223,
  10.79.161.239,
  10.79.173.142,
  10.79.166.232,
  10.79.169.58,
  10.79.167.91
],
metadata: [
  {
    Primary: true,
    PrivateIpAddress: "10.79.164.78"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.161.141"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.165.42"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.169.58"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.161.239"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.166.232"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.173.142"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.167.91"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.173.187"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.163.21"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.166.223"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.162.143"
  }{
    Primary: false,
    PrivateIpAddress: "10.79.161.21"
  }
]
```

seems like the IP pool on ENI and the IP pool in data store has a drift. What could possibly cause this issue?

**Comment by jayanthvn** [comment]
@duxing -

> why pod ip being deleted is 10.79.169.100 where the ip being deleted from datastore is 10.79.162.37? there's no other concurrent delete request in the span I post here. I'd like to know if these 2 ips are supposed to be the same.

The pod IP deleted and IP freed from DS need not be same. 

During reconciliation to maintain the warm pool we can pick any free IP - https://github.com/aws/amazon-vpc-cni-k8s/blob/e864016e80cb311bb2f8fc4338ef79e9b3a53e47/pkg/ipamd/datastore/data_store.go#L1404

The 2nd issue can happen because of IMDS sync. Which instance type you are using?

**Comment by duxing** [comment]
@jayanthvn thx for the quick response!

> The pod IP deleted and IP freed from DS need not be same.
 
that makes sense. the pod IPs are a subset of the IPs in DS.

> The 2nd issue can happen because of IMDS sync. Which instance type you are using?

I'm using `r5.8xlarge`: 
> instance type is r5.8xlarge: 30 ips per ENI, 8 max ENI. 

seems like the out-of-sync is somewhat expected. what would be the impact of that? would that get to a state that affects networking for 1 or more pods on the node?

**Comment by duxing** [comment]
@jayanthvn a follow-up question on 1 of the error logs:
after upgrading from `v1.5.3` to `v1.9.3`(limited by the version of `k8s` I'm using), I noticed this error log volume increased a lot:
> Error finding unassigned IPs for ENI <eni-id>

when I looked at the context for this log, it seems like a surefire log right after a normal(and successful) pod IP deletion request (it fires after the delete request is received and processed by `ipamd` but before the rpc call to `plugin`). There are other paths where the related function can be called but looking at the error volume alone, I would not know exactly how many of the errors I saw are coming from the "too-soon" problem during a normal deletion flow.

is there a way to bypass the issue and reduce the error volume to reflect only actual issues?

**Comment by jayanthvn** [comment]
@duxing - When the metadata goes out of sync, we try to reconcile with EC2 and confirm if the datastore is still valid.

https://github.com/aws/amazon-vpc-cni-k8s/blob/v1.10.0/pkg/ipamd/ipamd.go#L1402

Did you directly move from 1.5.3 to 1.9.3?

**Comment by duxing** [comment]
@jayanthvn thx for the explanation.
I did the upgrade directly from `1.5.3` to `1.9.3` without running into issues (during the upgrade)

however, 2 weeks after the upgrade, I noticed the same networking failure again:
 - instance networking is fine (`ssh`able, `curl`s work on the host
 - pods on the node are having networking issue: connection refused, connection timeout, etc
 - pods that are being `SIGTERM`'ed stuck in `Terminating` state
 - pods that are newly created fails initialization


from what I see, it feels like an issue with the internal state of `aws-cni`. it feels like a leaking issue: doesn't happen right away and takes time to get tipped over. 

I also have another cluster running the same version of k8s and aws-cni but handles a different type of workloads. the same issue was never observed on this cluster before or after the upgrade.

would love to know if the symptoms described reminds you of anything and if there are any suggestions to try. `1.9.3` is the latest version of `aws-cni` that's compatible with the k8s version we're running (having difficulty upgrading that)

**Comment by duxing** [comment]
@jayanthvn if I understand your response correctly, it's possible for the datastore(from `ipamd`) and the actual IPs from the ENIs to be out-of-sync, and this is addressed by reconciling the differences.

what needs to happen to generate a behavior that I noticed? I'd like to know if there are way to bypass it on the current version using configurations

**Comment by duxing** [comment]
@jayanthvn did further research on the error log I saw and found this issue(#282 )

seems like it's been closed without a definitive solution. I noticed this issue on version 1.5.3 and noticed it again on version 1.9.3

this port is bound to the `aws-k8s-agent` process. do you know which part of the agent listens on port `50051`? is there any native metrics we can turn to to verify things like how often does `aws-k8s-agent` crash?

**Comment by duxing** [comment]
more contex and findings that may be helpful:
1. the nodes I mentioned run into the issue have been running for days / weeks prior to the issue happening.
2. I suspect the "connection refuse" is caused by grpc server in `ipamd` crashing/getting terminated. I was able to spot the  "_Serving RPC Handler version_" log seconds after the wave of "connection refuse" error log from cni plugin.

then I looked at the logs from the window when the grpc server restart happened. prior to a "_Starting L-IPAMD_" log is a window of 2mins without any logs to `ipamd.log`, and the last few logs from 2mins ago are:
 - `DEBUG`: "getDeletableENI: found a deletable ENI <eni_id>"
 - `INFO`: "RemoveUnusedENIFromStore <eni-id>: IP/Prefix address pool stats: free 0 addresses, total: 53, assigned: 40, total prefixes: 0"
 - `DEBUG`: "Start freeing ENI <eni-id>"

the last status debug log before the stop of the `ipamd.log` is "_Current warm IP stats: target: 10, total: 53, assigned: 40, available: 13, short: 0, over 3_"

after the ipamd restart the first status log is "_Current warm IP stats: target: 10, total: 54, assigned: 5, available: 49, short: 0, over 39_"

then ipamd started deleting a bunch of IPs from the datastore and unassign IPs from the ENI.

questions for @jayanthvn :
1. is it normal to observe `ipamd` shutdown (silent) weeks after it's launched? I checked for the "Serving RPC Handler version"  log across all the nodes in my cluster in the past week and only saw entries from a few nodes. I also checked for "Received shutdown signal" log and found nothing. 
2. if `ipamd` restart is somewhat expected, can you help me explain the discrepancy between the last status log before shutdown and the 1st status log after restart? I think the discrepancy here is what caused IPs to be released and routing table rules to be cleared.  the shutdown and restart of `ipamd` is only 2mins apart, and there's not that many pod evictions during that 2 mins.

**Comment by achevuru** [comment]
@duxing It appears that you're running in to resource exhaustion/leaks (sockets & thread limit breaches; CPU/Mem) on the nodes. We've observed similar behavior/symptoms ( "connection refuse" --> leading to probe failures-> leading to restarts) when the nodes run in to sockets/threads limit breaches. In addition, you're running in to it on nodes after they're active for weeks without issues - so I would monitor the resource usage pattern on these nodes and check if some application running on these nodes is leaking resources.

Also, if you're observing IPAMD releasing IP addresses that are assigned to some pods on the nodes then check the memory consumption on the affected nodes.

**Comment by duxing** [comment]
hi @achevuru , thanks for the quick response! 

I've looked into this issue for a couple weeks and I agree the symptoms looks like a resource leak. 
The part I didn't mention in my previous msgs is I've checked the resource (primarily memory and CPU) usage pattern on affected nodes and didn't identify a smoking gun as the root cause. both CPU and memory have a lot of room when the issue is triggered. 

I'd like to understand the internals a bit more, if possible. Is `ipamd` shutting down silently an expected behavior? if the root cause is resource exhaustion/leak, I'd expect a related error/exception log, but I didn't see anything in the error logs.
in the sequence I described, I'd like to know which step is the 1st thing that deviates from what's normal and figure out a way to explain that.

**Comment by achevuru** [comment]
IPAMD is not expected shut down silently. When you exhaust sockets/thread limits for example - kubelet's liveness probe/readiness probe checks will fail for VPC CNI pods and that leads to IPAMD restarts. Which error logs are you referring to? IPAMD? You'll not see anything in IPAMD logs - you'll just see it restart randomly because of probe failures.

You can check kernel logs for any related errors and/or kubelet logs for pod readiness/liveness probe failures.

**Comment by duxing** [comment]
If I understand you correctly: if the healthiness/readiness probe is failing due to host resource exhaustion, the failure should not be specific to aws-cni right?

Another question on this theory: if readiness/healthiness probe fails, k8s should send SIGTERM/SIGKILL, which should b²e captured by the shutdown handler in aws-cni and log a line for graceful shutdown. I didn't see that line of log on my end, lead me to think it's crashing silently.

I've checked system resources usage again (memory, cpu, networking, disk) and the usage level is healthy, and stayed stable for the past week 

Let me check the k8s audit log to see if they's any logs supporting this theory and I'll get back to you on this.

Thanks again!

**Comment by duxing** [comment]
I'm seeing this very often now:
```
containerStatuses:
  - containerID: docker://<hash>
    image: 602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.9.3
    imageID: docker-pullable://602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni@sha256:<hash>
    lastState:
      terminated:
        containerID: docker://<hash>
        exitCode: 137
        finishedAt: "2023-05-03T03:11:28Z"
        reason: Error
        startedAt: "2023-05-03T02:00:18Z"
    name: aws-node
    ready: false
    restartCount: 1
    state:
      running:
        startedAt: "2023-05-03T03:11:29Z"
```

**Comment by duxing** [comment]
seems like this is caused by this error:
> Initialization failure: datastore: IP already in data store

**Comment by jayanthvn** [comment]
Do you see duplicate IPs in ipamd.json ? Since you moved from 1.5.x and there were issues with duplicate IPs for pods (https://github.com/aws/amazon-vpc-cni-k8s/issues/711) which was fixed partially in 1.6.4 and all the fixes went in 1.7.2. One possibility is pre-upgrade there were duplicate IP for pods, we need to check the ipamd.logs with 1.5.x. You seemed to be hitting this - https://github.com/aws/amazon-vpc-cni-k8s/blob/release-1.9/pkg/ipamd/datastore/data_store.go#L403

**Comment by duxing** [comment]
hi @jayanthvn thanks for the quick response!
I performed the upgrade about 3 weeks ago and no issues were observed until 3 days ago. 

ran into the same networking issue and thanks to the more structured logs in v1.9.3 I was able to figured out how that happened: `ipamd` was restarted and the state of IPs changes after restart (previous occupied IPs were detected as available IPs), resulting a wave of unassigning IPs that were occupied by pods.

the above was done with setting `WARM_IP_TARGET` to `10` and **NOT** setting `MINIMUM_IP_TARGET`. I attempted to set `MINIMUM_IP_TARGET` to match my average number of pods to see if it would help preserve the IPs in case the same issue happens, and this change has caused many `aws-node` pod stuck in `Terminating` state  during peak traffic hours. 

> Initialization failure: datastore: IP already in data store

is noticed for the first time today while I was looking into why `aws-node` pod failed to stand up after being restarted by `k8s`.  I still don't understand why `ipamd` will be restarted (surfaced as nothing listening on port 50051) when the host has 100+GB free memory and connection count is around 200

**Comment by duxing** [comment]
@jdn5126 thx for the response!

unfortunately `/var/run/aws-node/ipam.json` is not persisted on our log query solution. I'll submit a request to stream that path as well.

> Are you able to reproduce this on a new node

yes it has happened many times on newly launched node. I believe the duplicated IPs come from cni getting restarted  by k8s due to failing `livenessprobe` and internal state may have data persistency issue while trying to maining `MINIMUM_IP_TARGET`

I agree that both versions are really old and I'm trying to stop
the bleeding to I can have time to work on an upgrade solution. 

I don't have to commit to a correct fix given the context. If possible, I'd like to get more understanding on how this happened (ideally with logs/metrics to verify) and what are the band-aid solutions that I can apply to keep it stable for the next couple months. Thanks in advance!

**Comment by duxing** [comment]
any suggestions on the resource specification for `aws-node`?
I noticed in various versions of the yaml definition, only `requests.cpu` is provided (ranging from `10m` to `25m` based on version) but no `requests.memory`, `limits.memory` or `limits.cpu` are specified.

while 20m CPU is sufficient for most of the time, I do see CPU usage spike to `1000m` occasionally. with some other workloads (on the same node) also don't specify limits.cpu, I wonder if the `livenessprobe` failed due to CPU starvation.

I've also tested setting all of the resource specifications:
```
resources:
  requests:
    cpu:"100m"
    memory: "512Mi"
  limits:
    cpu:"100m"
    memory: "512Mi"
```
seems like this would have been sufficient resources but under this configuration I'm seeing more restarts of the cni and more networking issues on the same cluster.

forgot to mention: when I use the default `resources` setting, I noticed 137 exit code from `aws-cni` container sometimes when the issue happened
